### ISSUE TITLE
inaccurate tflite model

### ISSUE THREAD

**Issue by dillon3344** [tensorflow/tensorflow#53278]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installation (pip package or built from source): 2.7.0
- TensorFlow library (version, if pip package or github SHA, if built from source):

### 2. Code

```
# h5 model (correct)
import tensorflow as tf
import numpy as np

data = np.ones((1, 320, 160, 6), dtype=np.float32)
model = tf.keras.models.load_model("model_float32.h5")
model.predict(data)
```

```
# tflite full integer quantization model (incorrect)
import tensorflow as tf
import numpy as np

interpreter = tf.lite.Interpreter(model_path="model_quant_uint8.tflite")
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

interpreter.set_tensor(input_details[0]['index'], np.ones((1, 320, 160, 6), dtype=np.uint8))
interpreter.invoke()

interpreter.get_tensor(output_details[0]['index'])
```

### 3. Failure after conversion
If the conversion is successful, but the generated model is wrong, then state what is wrong:

- Model produces wrong results and/or has lesser accuracy.

### 4. (optional) RNN conversion support
If converting TF RNN to TFLite fused RNN ops, please prefix [RNN] in the title.

### 5. (optional) Any other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

links to models:
https://www.dropbox.com/s/pssd04thvgndo74/model_float32.h5?dl=0
https://www.dropbox.com/s/fpititzzwcsdmi2/model_float32.tflite?dl=0
https://www.dropbox.com/s/k88obxfdb3uowov/model_quant_uint8.tflite?dl=0

**Comment by tilakrayal** [comment]
@dillon3344 ,
Can you please provide the results of the model to confirm the issue.It helps to analyse and debug the issue.Thanks!

**Comment by dillon3344** [comment]
sure, here you go

model_float32.h5:
array([[-7.0198089e-02,  3.5972762e-01, -2.1505730e-01,  3.5191351e-01,
         8.9106321e-02,  7.9619581e-01, -1.8258725e+00, -1.8856851e+00,
        -2.0833955e+00, -3.4985578e+00, -4.3255777e+00, -3.1656742e+00,
         9.5930970e-01,  2.2216064e-01,  1.6680446e-02,  1.6571397e-01,
         4.5229942e-02, -3.8629866e+00, -5.0174055e+00, -4.7954006e+00,
        -6.6687365e+00,  9.8105502e-01,  4.9424553e-01, -1.2304291e-02,
         1.4696333e-01,  4.4353329e-02, -3.7876964e+00, -4.8331466e+00,
        -4.8142977e+00, -6.5505095e+00,  9.6800566e-01,  3.4754962e-02,
         4.0744931e-02,  8.8596672e-02,  2.5172849e-05,  7.3022842e-03,
         6.9045961e-02,  5.5118501e-03,  9.9129879e-01]], dtype=float32)

model_quant_uint8.tflite:
array([[204, 223, 217, 214, 223, 243, 202, 201, 197, 144, 130, 161, 239,
        208, 221, 225, 223, 168, 115, 115,  21, 249, 214, 222, 225, 223,
        169, 113, 113,  26, 248, 234, 232, 236, 223, 239, 253, 231, 224]],
      dtype=uint8)

**Comment by tilakrayal** [comment]
@sanatmpa1 ,
I was able to reproduce the issue in tf v2.5, v2.7 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/eb873ac1fa6dc6175bcc650ed5000867/53278.ipynb).

**Comment by dillon3344** [comment]
@sanatmpa1 any update?

**Comment by dillon3344** [comment]
@sachinprasadhs any update?

**Comment by sachinprasadhs** [comment]
Can you try with the post training quantization by referring [this](https://www.tensorflow.org/lite/performance/post_training_quantization) documentation, also quantization is subjected to loss of accuracy, you can check if the difference in accuracy is in acceptable range [here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/evaluation/tasks).

**Comment by xsacha** [comment]
I've had this issue on all models. Even Resnet 50.
Is there a version where this quantization works?

**Comment by dillon3344** [comment]
@sachinprasadhs i have applied post training quantization. Is there a way to compare the outputs of each layer with the known good model to find where the problem is?

**Comment by sachinprasadhs** [comment]
You can compare the model with the final outputs using details mentioned [here](https://www.tensorflow.org/lite/guide/faq#how_do_i_test_that_a_tensorflow_lite_model_behaves_the_same_as_the_original_tensorflow_model), not at the layer level though.

**Comment by xsacha** [comment]
I used the quantize debugger to get the op-by-op accuracy you see above. It takes a bit longer. There's tutorials for how to set it up.

**Comment by xsacha** [comment]
Issue is definitely not stale.
Any update?

**Comment by xsacha** [comment]
Thanks for reopening this.
I tested yet another model that exhibits this same issue.

I have the Netron graph for before (Keras) and after (Tensorflow Lite int8):
![image](https://user-images.githubusercontent.com/61218/151966818-f8a72458-7c12-4c33-9e78-7864dd15c305.png)


I'm seriously confused how everyone else has this working because I'm yet to find a working model.

**Comment by xsacha** [comment]
Any updates on this? How it not be working for so long?
I just tried v2.9 and quantisation still does not work.

**Comment by mohantym** [comment]
@dillon3344 @xsacha !
I was able to replicate this issue in [2.11](https://colab.sandbox.google.com/gist/mohantym/5e4f5d495d7d69f9fa9844a1e74863a0/53278.ipynb#scrollTo=nGB_uJH8Ea24) and [nightly version](https://colab.sandbox.google.com/gist/mohantym/02b3d7a5ac87ff039fde45fcd17b096b/53278.ipynb#scrollTo=drFEpPKrGFXG)(2.12dev). Just quick FYI - Suspected layers can be disabled using [tfl.quantization_debugger ](https://www.tensorflow.org/lite/performance/quantization_debugger)now. 

Thank you!

**Comment by pjpratik** [comment]
Hi @dillon3344 

The model provides the output as `int8_value` which needs to be converted into `real_value` using quantization parameters, scale and zero point.

$$ real value = (int8 value - zero point) \times scale $$

Please find the [gist](https://colab.sandbox.google.com/gist/pjpratik/fca47ec6fdf6f8ce3d4e9a4efaf7a424/53278.ipynb) for the same.

Thanks.
